src/react-vite/08-stateful-hooks/stateful-hooks.md

```diff
@@ -2,51 +2,117 @@
+
+## Objective: Manage component state using hooks
+
+React provides hooks that simplify managing state in a component. We will use
+some of these hooks to store the user's selections for state and city.
+
+> Note: Hooks must always appear in the body of the component **before** any
+> statements that might change the component's behavior. Place hooks at the top
+> of a component's body.
+
+### useState
+
+We can store state that persists through component rendering with the `useState`
+hook. You can set the initial state value when the component **first** renders
+by providing the value as an argument to the hook. _If you do not provide a
+value the initial state value will be `undefined`._
+
+This example shows a `useState` hook being set with an initial value from an
+`initialText` prop.
+
+```jsx
+const FancyInput: React.FC<Props> = ({ initialText }) => {
+  const [value, setValue] = useState(initialText)
+}
+```
+
+As you can see in the previous example, `useState` returns an array with two
+elements: the first is the current value of the hook, and the second is a setter
+function that is used to update the state value. In the following code, the
+value and setter are being used to update changes to an input element.
+
+```jsx
+const FancyInput: React.FC<Props> = ({ initialText }) => {
+  const [value, setValue] = useState(initialText)
+  return <input onChange={(evt) => setValue(evt.target.value)} value={value} /> 
+}
+```
+
+Every time a `useState`'s setter is invoked with a new value, React compares the
+new value with the current value. If the values are the same, nothing happens;
+**if the values are different, React will rerender the component** so the new
+value can be used to update the component. In the example above, when the user
+types or changes the value of the input, the `FancyInput` component is rendered
+again, and the input is updated with the current value.
+
+> Watch out! To determine if a state value has changed React compares value
+> types like `object` and array by instance, not by their contents. Other types,
+> like number, string, and boolean, are compared by their value.
+
+### useId
+
+One of the newer hooks that React includes is `useId`. Since the value of every
+`id` attribute in an HTML document must be unique, this hook is useful in
+creating a unique identifier string that can be used as the value of `id` props.
+
+```jsx
+const FancyInput: React.FC<Props> = ({ initialText }) => {
+  const [value, setValue] = useState(initialText)
+  const id = useId();
+  return (
+    <label htmlFor={id}>Name:</label>
+    <input id={id} onChange={(evt) => setValue(evt.target.value)} value={value} />
+  )
+}
+```
 
-## Objective 1
-
-TODO
-
-### Key concepts
+### Setup
 
-TODO
+<span style="color:red">I think we should add the code related to the state
+object arrays, form, and the selects during setup. Then the exercise will
+involve adding the hooks and wiring state and the setters up to the code.</span>
```

Comment:
Yeah, the folders I have are just the solutions, not the problem/setup. Sorry I had run out of time to do that and was going to but you beat me to starting on this section. 😃

For this exercise, I think the only thing folks should be doing is to add anything that is/touches `stateId`, `cityId`, `state`, and `city`. So, the imports and the HTML should all be there when they start, then they should add the state variables to the top of the component and hook them up to the `htmlFor`, `onChange`, etc.

I think maybe we should event split the exercise in two? Maybe focus on `useState` first and then `useId` afterwards, or vice versa? Then we can split up the explanations to so each objective really focuses on one thing. That should work better for the training.

Reply:
done